crates/red_knot_python_semantic/src/types.rs

```diff
@@ -4542,6 +4546,77 @@ impl<'db> GeneralCallableType<'db> {
             .is_some_and(|return_type| return_type.is_fully_static(db))
     }
 
+    /// Return `true` if `self` represents the exact same set of possible runtime objects as `other`.
+    pub(crate) fn is_equivalent_to(self, db: &'db dyn Db, other: Self) -> bool {
+        let self_signature = self.signature(db);
+        let other_signature = other.signature(db);
+
+        let self_parameters = self_signature.parameters();
+        let other_parameters = other_signature.parameters();
+
+        if self_parameters.len() != other_parameters.len() {
+            return false;
+        }
+
+        if self_parameters.is_gradual() || other_parameters.is_gradual() {
+            return false;
+        }
+
+        // Check equivalence relationship between two optional types. If either of them is `None`,
+        // then it is not a fully static type which means it's not equivalent either.
```

Comment:
Ultimately we will also need `is_gradual_equivalent_to`, in which `None` vs `None` should be considered equivalent. Similar to what we previously discussed for `is_subtype_of` vs `is_assignable_to`, it may end up being simplest to rename this method to `is_gradual_equivalent_to` and ensure it handles all types correctly, and in the `is_equivalent_to_` case it will only be called with fully-static callable types.

Reply:
We already have `is_gradual_equivalent_to` but I understand what you mean. I'm going to first explore whether we can maintain the equivalence relation at the type level and then see if this needs to be simplified.

Reply:
Hmm, taking a look at the current implementation of `is_gradual_equivalent_to`, I think it's incorrect because it only looks at the annotated type for the parameters and not any other properties like name, default, and kind. This means that the following two callable types are gradual equivalent:

```py
def f1(a): ...
def f2(b): ...
```

I think I'll fix this as a follow-up which will then also include the simplification because it needs to special case the `...` gradual form.